numba/tests/test_array_attr.py

```diff
@@ -228,13 +228,50 @@ def test_size_after_slicing(self):
             self.assertEqual(pyfunc(arr, i), cfunc(arr, i))
 
 
-class TestArrayCTypes(MemoryLeakMixin, unittest.TestCase):
+class TestArrayCTypes(MemoryLeakMixin, TestCase):
     def test_array_ctypes_data(self):
         pyfunc = array_ctypes_data
         cfunc = njit(pyfunc)
         arr = np.arange(3)
         self.assertEqual(pyfunc(arr), cfunc(arr))
 
+    def test_array_ctypes_ref_error_in_parallel(self):
+        # Issue #2887
+        from ctypes import CFUNCTYPE, c_void_p, c_int32, c_double, c_bool
+
+        @CFUNCTYPE(c_bool, c_void_p, c_int32, c_void_p)
+        def callback(inptr, size, outptr):
+            # A ctypes callback that manipulate the incoming pointers.
+            try:
+                inbuf = (c_double * size).from_address(inptr)
+                outbuf = (c_double * 1).from_address(outptr)
+                a = np.ndarray(size, buffer=inbuf, dtype=np.float64)
+                b = np.ndarray(1, buffer=outbuf, dtype=np.float64)
+                b[0] = (a + a.size)[0]
+                return True
+            except:
+                import traceback
+                traceback.print_exception()
+                return False
+
+
+        # parallel=True is required to reproduce the error.
+        @njit(parallel=True)
+        def foo(size):
+            arr = np.ones(size)
+            out = np.empty(1)
+            # Exercise array.ctypes
+            inct = arr.ctypes
+            outct = out.ctypes
+            # The reference to `arr` is dead by now
```

Comment:
`out` ref is also dead?

Reply:
`out` ref is not dead.  It's being returned in the end.

Reply:
oh yes! oops!